### PR TITLE
add typing overloads so msg.fail with exits=True/1 has the NoReturn type like sys.exit

### PR DESCRIPTION
**Pull request by kabirkhan**
We use wasabi in all our libraries but with Pyright, we're unable to properly use the `exits=True/1` parameter to indicate the rest of the code in the scope won't execute. So we end up doing 

```python
msg.fail("some error")
sys.exit(1)
``` 

instead of 

```python
msg.fail("some error", exits=1)
```

This uses typing overloads to the `Printer.fail` method so pyright knows what to do when `exits=True` or `exits=1`

### CODE REVIEW

**Comment by adrianeboyd** [comment]
I don't think you need to change the requirements for this, `Literal` is already included in `compat` and I think we can assume that users have `NoReturn` with python 3.6. (If at all possible I'd like to plan to be able to drop `typing_extensions` as a requirement in the future.)

**Comment by kabirkhan** [comment]
Ok sounds good! I tend to just use `typing_extensions` when using types I haven't before without actually researching which python version they're introduced in (bad habit for sure). 

Thanks! reverted the reqs change here.

File: wasabi/printer.py [hunk 1]
```diff
@@ -92,14 +93,36 @@ def good(
...
+    @overload
+    def fail(
+        self,
+        title: Any = "",
+        text: Any = "",
+        show: bool = True,
+        spaced: bool = False,
+        exits: Literal[1, True] = True,
+    ) -> NoReturn:
+        ...
```
Comment (adrianeboyd) [inline wasabi/printer.py:116]: I think all the defaults here can be `...` instead?

File: wasabi/printer.py [hunk 2]
```diff
@@ -92,14 +93,36 @@ def good(
...
+    @overload
     def fail(
         self,
         title: Any = "",
         text: Any = "",
         show: bool = True,
         spaced: bool = False,
-        exits: Optional[int] = None,
+        exits: Optional[Literal[0, False]] = None,
     ):
```
Comment (adrianeboyd) [inline wasabi/printer.py:104]:
```suggestion
    ): -> Union[str, None]
```

Comment (kabirkhan) [reply wasabi/printer.py:104]: So for some reason, adding this type annotation gives a mypy error:

```
Overloaded function signatures 1 and 2 overlap with incompatible return types  [misc]
```

and I'm not sure why. If I leave that off, the error is not raised and type annotations work like expected still when calling the `Printer.fail` method

File: wasabi/printer.py (hunk 1 above)
Comment (kabirkhan) [reply wasabi/printer.py:116]: Unresolving these and putting these back to regular defaults. I know this is not super helpful in Mypy but if you're using VSCode autocomplete (and I assume other pyright editors) you see the `...` instead of the actual defaults here.

Comment (adrianeboyd) [reply wasabi/printer.py:116]: Does pyright detect if the defaults get out of sync?

I tested with mypy and I could make all the defaults different with no issues found by mypy, while only the defaults from the main method are used in practice.

My guess is that vscode would happily show you completely incorrect defaults in this case?

I don't really mean to hold up this particular (relatively minor) change, it's just very frustrating in general.

**Comment by adrianeboyd** [comment]
Can you double-check this with spacy + mypy? I'm getting a new mypy error for spacy with this change:

```
spacy/cli/project/run.py:104: error: Argument "exits" to "fail" of "Printer" has incompatible type "Optional[int]"; expected "Optional[Literal[0, False]]"  [arg-type]
```

**Comment by adrianeboyd** [comment]
Sorry, I think it's okay to leave the defaults in here if it's helpful in vscode. It's not like this code changes very often. I'm just annoyed at the whole situation. (We have other cases with overloads in spacy where we have `...` and I thought it was the intended way to do it.)

I tested again in a clean venv with spacy and I still get the same mypy error. Let me try again in the spacy CI to double-check. It might be possible that it depends on the python version?

**Comment by adrianeboyd** [comment]
So as best I can tell something changed between python 3.8.10 and 3.8.16 and with a newer version of python 3.8 it's also fine on my end. How strange...
